### PR TITLE
Introduce HalfGraphType

### PR DESCRIPTION
**Pull request by sungam3r**
https://devblogs.microsoft.com/dotnet/introducing-the-half-type/

### CODE REVIEW

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 1]
```diff
@@ -155,11 +165,20 @@ public void does_not_coerce_string(Type graphType)
     [InlineData(typeof(BigIntGraphType), -1E+25)]
     [InlineData(typeof(BigIntGraphType), 0)]
     [InlineData(typeof(BigIntGraphType), 1E+25)]
+#if NET5_0_OR_GREATER
+    [InlineData(typeof(HalfGraphType), -65504)]
+    [InlineData(typeof(HalfGraphType), 65504)]
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:170]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65504)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 2]
```diff
@@ -228,17 +247,33 @@ public void parseValue_ok(Type graphType, object value)
     [InlineData(typeof(BigIntGraphType), -1E+25)]
     [InlineData(typeof(BigIntGraphType), 0)]
     [InlineData(typeof(BigIntGraphType), 1E+25)]
+#if NET5_0_OR_GREATER
+    [InlineData(typeof(HalfGraphType), -65500)]
+    [InlineData(typeof(HalfGraphType), 65500)]
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:252]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65500)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 3]
```diff
@@ -270,18 +305,34 @@ public void parseValue_from_newtonsoft_ok(Type graphType, object value)
...
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:310]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65500)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 4]
```diff
@@ -311,11 +362,20 @@ public void parseValue_from_system_text_json_ok(Type graphType, object value)
     [InlineData(typeof(BigIntGraphType), -1E+25)]
     [InlineData(typeof(BigIntGraphType), 0)]
     [InlineData(typeof(BigIntGraphType), 1E+25)]
+#if NET5_0_OR_GREATER
+    [InlineData(typeof(HalfGraphType), -65504)]
+    [InlineData(typeof(HalfGraphType), 65504)]
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:367]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65504)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 5]
```diff
@@ -370,11 +430,20 @@ public void parseLiteral_ok(Type graphType, object value)
...
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:435]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65504)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 6]
```diff
@@ -440,11 +509,20 @@ public void serialize_ok(Type graphType, object value)
...
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:514]:
```suggestion
    [InlineData(typeof(HalfGraphType), 0)]
    [InlineData(typeof(HalfGraphType), 65504)]
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 7]
```diff
@@ -728,6 +822,12 @@ public void parseValue_out_of_range(Type graphType, object value)
             {
                 Should.Throw<InvalidOperationException>(() => g.ParseValue(converted));
             }
+#if NET5_0_OR_GREATER
+            else if (graphType == typeof(HalfGraphType))
+            {
+                Should.Throw<InvalidOperationException>(() => g.Serialize(converted));
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:828]:
```suggestion
                Should.Throw<InvalidOperationException>(() => g.ParseValue(converted));
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 8]
```diff
@@ -622,6 +704,10 @@ public void parseValue_other_fail(Type graphType, object value)
     [InlineData(typeof(LongGraphType), 1.5)]
     [InlineData(typeof(ULongGraphType), 1.5)]
     [InlineData(typeof(BigIntGraphType), 1.5)]
+#if NET5_0_OR_GREATER
+    [InlineData(typeof(HalfGraphType), 1000000)]
+    [InlineData(typeof(HalfGraphType), -1000000)]
+#endif
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:710]:
```suggestion
#if NET5_0_OR_GREATER
    [InlineData(typeof(HalfGraphType), false)]
#endif
```
the "other_fail" tests are for testing invalid types passed to the parseliteral method; out of range values are handled by parseLiteral_out_of_range

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 9]
```diff
@@ -667,6 +753,10 @@ public void parseLiteral_other_fail(Type graphType, object value)
...
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:759]:
```suggestion
#if NET5_0_OR_GREATER
    [InlineData(typeof(HalfGraphType), false)]
#endif
```

File: src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs [hunk 10]
```diff
@@ -587,6 +665,10 @@ public void serialize_other_ok(Type graphType, object value, object serialized)
...
```
Comment (Shane32) [inline src/GraphQL.Tests/Types/AllScalarGraphTypeTests.cs:671]:
```suggestion
#if NET5_0_OR_GREATER
    [InlineData(typeof(HalfGraphType), false)]
#endif
```

File: src/GraphQL/Types/Scalars/HalfGraphType.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// <inheritdoc/>
+    public override object? ParseValue(object? value) => value switch
+    {
+        double db => NotInfinity(checked((Half)db)),
+        int i => NotInfinity(checked((Half)i)),
+        null => null,
+        float f => NotInfinity(checked((Half)f)),
+        decimal d => NotInfinity(checked((Half)(double)d)),
+        sbyte sb => checked((Half)sb),
```
Comment (Shane32) [inline src/GraphQL/Types/Scalars/HalfGraphType.cs:56]:
```suggestion
        Half h => value,
        sbyte sb => checked((Half)sb),
```

File: src/GraphQL/Types/Scalars/HalfGraphType.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    private object NotInfinity(Half value)
+    {
+        return Half.IsInfinity(value) ? ThrowValueConversionError(value) : value;
+    }
+
+    /// <inheritdoc/>
+    public override object? ParseValue(object? value) => value switch
+    {
+        double db => NotInfinity(checked((Half)db)),
+        int i => NotInfinity(checked((Half)i)),
```
Comment (Shane32) [inline src/GraphQL/Types/Scalars/HalfGraphType.cs:52]:
```suggestion
        int i => checked((Half)i),
```
Integer conversions should throw if out of range

File: src/GraphQL/Types/Scalars/HalfGraphType.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        float f => NotInfinity(checked((Half)f)),
+        decimal d => NotInfinity(checked((Half)(double)d)),
+        sbyte sb => checked((Half)sb),
+        byte b => checked((Half)b),
+        short s => NotInfinity(checked((Half)s)),
+        ushort us => NotInfinity(checked((Half)us)),
+        uint ui => NotInfinity(checked((Half)ui)),
+        long l => NotInfinity(checked((Half)l)),
+        ulong ul => NotInfinity(checked((Half)ul)),
+        BigInteger bi => NotInfinity(checked((Half)(double)bi)),
```
Comment (Shane32) [inline src/GraphQL/Types/Scalars/HalfGraphType.cs:63]:
```suggestion
        uint ui => checked((Half)ui),
        long l => checked((Half)l),
        ulong ul => checked((Half)ul),
        BigInteger bi => checked((Half)(double)bi),
```

File: src/GraphQL/Types/Scalars/HalfGraphType.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        double db => NotInfinity(checked((Half)db)),
+        int i => NotInfinity(checked((Half)i)),
+        null => null,
+        float f => NotInfinity(checked((Half)f)),
+        decimal d => NotInfinity(checked((Half)(double)d)),
+        sbyte sb => checked((Half)sb),
+        byte b => checked((Half)b),
+        short s => NotInfinity(checked((Half)s)),
+        ushort us => NotInfinity(checked((Half)us)),
```
Comment (Shane32) [inline src/GraphQL/Types/Scalars/HalfGraphType.cs:59]:
```suggestion
        short s => checked((Half)s),
        ushort us => checked((Half)us),
```

**Review by Shane32 — COMMENTED** [review]
We don't have a scalar for single-precision floating-point values.  Perhaps we should do that also.  And there's the new Int128 type from .NET 7.  And as of now, all of the ParseValue functions for each numeric scalar will cast any numeric value to the specified type.  To keep this behavior, we may want to add `Half` conversions to the other types.

**Comment by sungam3r** [comment]
>  I probably wouldn't as I believe .NET requires an explicit cast to convert Half to anything else. (??)

I wouldn't as well. Yes, `Half` type is very limited now in APIs.

File: src/GraphQL/Types/Scalars/HalfGraphType.cs (hunk 11 above)
Comment (sungam3r) [reply src/GraphQL/Types/Scalars/HalfGraphType.cs:56]: oops)

File: src/GraphQL/Types/Scalars/HalfGraphType.cs (hunk 12 above)
Comment (sungam3r) [reply src/GraphQL/Types/Scalars/HalfGraphType.cs]: изображение

I return `NonInfinity` back.

Comment (Shane32) [reply src/GraphQL/Types/Scalars/HalfGraphType.cs]: Ok

**Review by Shane32 — APPROVED** [review]

**Pull request merged**